pkg/controller/constraint/constraint_controller.go

```diff
@@ -47,85 +46,103 @@ var (
 )
 
 const (
-	finalizerName = "finalizers.gatekeeper.sh/constraint"
+	finalizerName    = "finalizers.gatekeeper.sh/constraint"
+	constraintsGroup = "constraints.gatekeeper.sh"
 )
 
 type Adder struct {
 	Opa              *opa.Client
 	ConstraintsCache *ConstraintsCache
+	WatchManager     *watch.Manager
+	ControllerSwitch *watch.ControllerSwitch
+	Events           <-chan event.GenericEvent
 }
 
-type ConstraintsCache struct {
-	mux   sync.RWMutex
-	cache map[string]tags
+func (a *Adder) InjectOpa(o *opa.Client) {
+	a.Opa = o
 }
 
-type tags struct {
-	enforcementAction util.EnforcementAction
-	status            metrics.Status
+func (a *Adder) InjectWatchManager(w *watch.Manager) {
+}
+
+func (a *Adder) InjectControllerSwitch(cs *watch.ControllerSwitch) {
+	a.ControllerSwitch = cs
 }
 
 // Add creates a new Constraint Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
 // and Start it when the Manager is Started.
-func (a *Adder) Add(mgr manager.Manager, gvk schema.GroupVersionKind, cs *watch.ControllerSwitch) error {
+func (a *Adder) Add(mgr manager.Manager) error {
 	reporter, err := newStatsReporter()
 	if err != nil {
 		log.Error(err, "StatsReporter could not start")
 		return err
 	}
 
-	r := newReconciler(mgr, gvk, a.Opa, cs, reporter, a.ConstraintsCache)
-	return add(mgr, r, gvk)
+	r := newReconciler(mgr, a.Opa, a.ControllerSwitch, reporter, a.ConstraintsCache)
+	return add(mgr, r, a.Events)
+}
+
+type ConstraintsCache struct {
+	mux   sync.RWMutex
+	cache map[string]tags
+}
+
+type tags struct {
+	enforcementAction util.EnforcementAction
+	status            metrics.Status
 }
 
 // newReconciler returns a new reconcile.Reconciler
 func newReconciler(
 	mgr manager.Manager,
-	gvk schema.GroupVersionKind,
 	opa *opa.Client,
 	cs *watch.ControllerSwitch,
 	reporter StatsReporter,
 	constraintsCache *ConstraintsCache) reconcile.Reconciler {
 	return &ReconcileConstraint{
-		Client:           mgr.GetClient(),
+		// Separate reader and writer because manager's default client bypasses the cache for unstructured resources.
```

Comment:
> Separate reader and writer because manager's default client bypasses the cache for unstructured resources.

Is there any documentations around this?

could there be any issues with `reader.Get` or `reader.List` when reading from the cache? e.g. data freshness

Reply:
I don't know that it's documented, but the code is here: https://github.com/kubernetes-sigs/controller-runtime/blob/v0.5.0/pkg/manager/manager.go#L328-L330

Regarding data freshness - we should be good. Although the watch.Manager does not block for cache sync when adding a new watch, if you do a `reader.Get` or `reader.List`, controller-runtime _will_ block on `WaitForCacheSync` here: https://github.com/open-policy-agent/gatekeeper/blob/5ff95d9bee9aa424f532974891b4385609afde9d/third_party/sigs.k8s.io/controller-runtime/pkg/dynamiccache/internal/informers_map.go#L203